### ISSUE TITLE
ctest of bl_test_unit failed on fedora s390x

### ISSUE THREAD

**Issue by topazus** [blend2d/blend2d#198]
After applying patch https://github.com/blend2d/blend2d/pull/197, on s390x, CMake build ok, bl_test_unit ctest failed.

<details>

```bash
+ /usr/bin/ctest --test-dir redhat-linux-build --output-on-failure --force-new-ctest-process -j2
Internal ctest changing into directory: /builddir/build/BUILD/blend2d-5a263ce51f3f880ee6c60f6345d18c3eccbe200f/redhat-linux-build
Test project /builddir/build/BUILD/blend2d-5a263ce51f3f880ee6c60f6345d18c3eccbe200f/redhat-linux-build
    Start 1: bl_test_unit
    Start 2: bl_test_fuzzer
1/4 Test #1: bl_test_unit .....................Subprocess aborted***Exception:   2.04 sec
Blend2D Unit Tests [use --help for command line options]
  Version    : 0.10.6
  Build Type : Release
  Compiled By: GCC 14.0

[[Priority=1]]
Running simd

[[Priority=2]]
Running math
  bl::Math::floor()
  bl::Math::ceil()
  bl::Math::trunc()
  bl::Math::round()
  bl::Math::floorToInt()
  bl::Math::ceilToInt()
  bl::Math::truncToInt()
  bl::Math::roundToInt()
  bl::Math::frac()
  bl::Math::isBetween0And1()
  bl::Math::quadRoots
  bl::Math::cubicRoots
Running support_algorithm
  bl::lowerBound()
  bl::quickSort() - Testing quickSort and insertionSort of predefined arrays
  bl::quickSort() - Testing quickSort and insertionSort of artificial arrays
  bl::quickSort() - Testing quickSort and insertingSort with unstable compare function
  bl::binarySearch() - Testing binary search
Running support_bitops
  bl::ParametrizedBitOps<bl::BitOrder::kLSB>::bitArrayFill
  bl::ParametrizedBitOps<bl::BitOrder::kMSB>::bitArrayFill
  bl::ParametrizedBitOps<bl::BitOrder::kLSB>::BitIterator<uint32_t>
  bl::ParametrizedBitOps<bl::BitOrder::kMSB>::BitIterator<uint32_t>
  bl::ParametrizedBitOps<bl::BitOrder::kLSB>::BitVectorIterator<uint32_t>
  bl::ParametrizedBitOps<bl::BitOrder::kMSB>::BitVectorIterator<uint32_t>
  bl::ParametrizedBitOps<bl::BitOrder::kLSB>::BitVectorFlipIterator<uint32_t>
  bl::ParametrizedBitOps<bl::BitOrder::kMSB>::BitVectorFlipIterator<uint32_t>
Running support_intops_alignment
  Testing bl::IntOps::isAligned()
  Testing bl::IntOps::isPowerOf2()
  Testing bl::IntOps::alignUp()
  Testing bl::IntOps::alignUpDiff()
  Testing bl::IntOps::alignUpPowerOf2()
Running support_intops_arithmetic
  Testing bl::IntOps::addOverflow()
  Testing bl::IntOps::subOverflow()
  Testing bl::IntOps::mulOverflow()
Running support_intops_bit_manipulation
  Testing bl::IntOps::shl() / IntOps::shr()
  Testing bl::IntOps::rol() / IntOps::ror()
  Testing bl::IntOps::clz()
  Testing bl::IntOps::ctz()
  Testing bl::IntOps::isBitMaskConsecutive()
Running support_intops_byteswap
  Testing bl::IntOps::byteSwap16()
  Testing bl::IntOps::byteSwap24()
  Testing bl::IntOps::byteSwap32()
  Testing bl::IntOps::byteSwap64()
Running support_intops_clamp
  Testing bl::IntOps::clamp()
  Testing bl::IntOps::clampToByte()
  Testing bl::IntOps::clampToWord()
Running support_intops_popcnt
  Testing bl::IntOps::PopCounter<uint32_t>
Running support_memops
  bl::MemOps - Read & Write
  bl::MemOps - Copy Forward
Running support_ptrops
  bl::PtrOps - Offset / Deoffset

[[Priority=3]]
Running arena_hashmap
  Inserting 10000 elements to HashTable
  Removing 10000 elements from HashTable and validating each operation
Running arena_list
  Append / Unlink
  Prepend / Unlink
  InsertAfter / InsertBefore
  PopFirst / Pop
Running arena_tree
  Inserting 2000 elements to the tree and validating each operation
  Removing 2000 elements from the tree and validating each operation
Running support_arenabitarray
  bl::ArenaBitArray::resize()
  bl::ArenaBitArray::fillBits() / clearBits()

[[Priority=4]]
Running thread_pool
  Trying to allocate very high number of threads that should fail
  Repeatedly acquiring / releasing 4 threads with a simple task
  [#0] Acquiring 4 threads from thread-pool
  [#0] Running 4 threads
  [#0] Waiting and releasing
  [#0] Thread 0x2aa39858400 running
  [#0] Thread 0x2aa39858800 running
  [#0] Thread 0x2aa39858a00 running
  [#0] Thread 0x2aa39858600 running
  [#0] Thread 0x2aa39858600 signaling to main thread
  [#1] Acquiring 4 threads from thread-pool
  [#1] Running 4 threads
  [#1] Thread 0x2aa39858400 running
  [#1] Thread 0x2aa39858800 running
  [#1] Waiting and releasing
  [#1] Thread 0x2aa39858600 running
  [#1] Thread 0x2aa39858a00 running
  [#1] Thread 0x2aa39858a00 signaling to main thread
  [#2] Acquiring 4 threads from thread-pool
  [#2] Running 4 threads
  [#2] Thread 0x2aa39858400 running
  [#2] Waiting and releasing
  [#2] Thread 0x2aa39858a00 running
  [#2] Thread 0x2aa39858600 running
  [#2] Thread 0x2aa39858800 running
  [#2] Thread 0x2aa39858a00 signaling to main thread
  [#3] Acquiring 4 threads from thread-pool
  [#3] Running 4 threads
  [#3] Thread 0x2aa39858400 running
  [#3] Waiting and releasing
  [#3] Thread 0x2aa39858800 running
  [#3] Thread 0x2aa39858a00 running
  [#3] Thread 0x2aa39858600 running
  [#3] Thread 0x2aa39858600 signaling to main thread
  [#4] Acquiring 4 threads from thread-pool
  [#4] Running 4 threads
  [#4] Thread 0x2aa39858400 running
  [#4] Waiting and releasing
  [#4] Thread 0x2aa39858a00 running
  [#4] Thread 0x2aa39858800 running
  [#4] Thread 0x2aa39858600 running
  [#4] Thread 0x2aa39858600 signaling to main thread
  [#5] Acquiring 4 threads from thread-pool
  [#5] Running 4 threads
  [#5] Thread 0x2aa39858400 running
  [#5] Waiting and releasing
  [#5] Thread 0x2aa39858800 running
  [#5] Thread 0x2aa39858600 running
  [#5] Thread 0x2aa39858a00 running
  [#5] Thread 0x2aa39858a00 signaling to main thread
  [#6] Acquiring 4 threads from thread-pool
  [#6] Running 4 threads
  [#6] Thread 0x2aa39858400 running
  [#6] Thread 0x2aa39858a00 running
  [#6] Thread 0x2aa39858800 running
  [#6] Waiting and releasing
  [#6] Thread 0x2aa39858600 running
  [#6] Thread 0x2aa39858600 signaling to main thread
  [#7] Acquiring 4 threads from thread-pool
  [#7] Running 4 threads
  [#7] Thread 0x2aa39858400 running
  [#7] Waiting and releasing
  [#7] Thread 0x2aa39858800 running
  [#7] Thread 0x2aa39858a00 running
  [#7] Thread 0x2aa39858600 running
  [#7] Thread 0x2aa39858600 signaling to main thread
  [#8] Acquiring 4 threads from thread-pool
  [#8] Running 4 threads
  [#8] Thread 0x2aa39858400 running
  [#8] Waiting and releasing
  [#8] Thread 0x2aa39858600 running
  [#8] Thread 0x2aa39858800 running
  [#8] Thread 0x2aa39858a00 running
  [#8] Thread 0x2aa39858a00 signaling to main thread
  [#9] Acquiring 4 threads from thread-pool
  [#9] Running 4 threads
  [#9] Thread 0x2aa39858400 running
  [#9] Waiting and releasing
  [#9] Thread 0x2aa39858a00 running
  [#9] Thread 0x2aa39858800 running
  [#9] Thread 0x2aa39858600 running
  [#9] Thread 0x2aa39858600 signaling to main thread
  Cleaning up
  Done

[[Priority=5]]
Running compression_crc32

[[Priority=6]]
Running random
  Testing whether the random 32-bit integer is the HI part of the 64-bit one
  Test whether returned double precision values satisfy [0..1) condition
    Random numbers at [0.0, 0.5): 500581 of 1000000
    Random numbers at [0.5, 1.0): 499419 of 1000000
Running tables
  Testing 'bl::commonTable.unpremultiplyRcp' correctness
Running unicode
Running unicode_io
  bl::Unicode::Utf8Reader
  bl::Unicode::Utf16Reader
  bl::Unicode::Utf32Reader
  bl::Unicode::Utf8Writer
  bl::Unicode::Utf16Writer
Running zero_allocator
  Memory alloc/release test - 50000 allocations
  Allocating zeroed memory...
    NumBlocks:        17
    UsedSize : 232305664 [Bytes]
    Reserved : 233832448 [Bytes]
    Overhead :      4928 [Bytes]
  Releasing zeroed memory...
    NumBlocks:         6
    UsedSize :         0 [Bytes]
    Reserved :  49283072 [Bytes]
    Overhead :      1232 [Bytes]
  Submitting manual cleanup...
    NumBlocks:         1
    UsedSize :         0 [Bytes]
    Reserved :   1048576 [Bytes]
    Overhead :        96 [Bytes]
  Allocating zeroed memory...
    NumBlocks:        17
    UsedSize : 231490560 [Bytes]
    Reserved : 233832448 [Bytes]
    Overhead :      4928 [Bytes]
  Shuffling...
  Releasing 50% blocks...
    NumBlocks:        17
    UsedSize : 115789824 [Bytes]
    Reserved : 233832448 [Bytes]
    Overhead :      4928 [Bytes]
  Allocating 50% blocks again...
    NumBlocks:        17
    UsedSize : 231491584 [Bytes]
    Reserved : 233832448 [Bytes]
    Overhead :      4928 [Bytes]
  Releasing zeroed memory...
    NumBlocks:         6
    UsedSize :         0 [Bytes]
    Reserved :  76546048 [Bytes]
    Overhead :      1648 [Bytes]

[[Priority=7]]
Running var_basic_types
  Verifying null value functionality
  Verifying bool value functionality
  Verifying int/uint value functionality
  Verifying float/double value functionality
  Checking bool/int/uint/double value conversions
Running var_styles
  Verifying BLRgba[32|64] value functionality
  Checking BLRgba[32|64] value conversions
  Checking BLGradient value functionality
  Checking BLPattern value functionality

[[Priority=8]]
Running array
  Basic functionality - BLArray<int>
  Basic functionality - BLArray<uint8_t>
  Basic functionality - BLArray<uint64_t>
  Basic functionality - C API
  External array
  String array
Running bitarray
  [SSO] Basic functionality
  [SSO] Using word quantities
  [Dynamic] Basic functionality
Running bitset
  Checking SSO BitSet basics
  Checking SSO BitSet ranges
  Checking SSO BitSet assignWords()
  Checking SSO BitSet chop()
  Checking SSO BitSet hasBitsInRange() & cardinalityInRange()
  Checking dynamic BitSet basics
  Checking dynamic BitSet addRange() & clearRange()
  Checking dynamic BitSet assignWords()
  Checking dynamic BitSet addWords() - small BitSet
  Checking dynamic BitSet addWords() - large BitSet
  Checking dynamic BitSet addWords() - consecutive
  Checking dynamic BitSet chop()
  Checking dynamic BitSet hasBitsInRange() & cardinalityInRange()
  Checking functionality of shrink() & optimize()
  Checking functionality of subsumes() & intersects()
  Checking functionality of BLBitSetBuilder
  Checking functionality of BLBitSetWordIterator
Running string
  SSO representation
  Assignment and comparison
  String manipulation
  String formatting
  String search
Running string_allocation_strategy
  Capacity increased from 11 to 15 [ImplSize=32]
  Capacity increased from 15 to 47 [ImplSize=64]
  Capacity increased from 47 to 111 [ImplSize=128]
  Capacity increased from 111 to 239 [ImplSize=256]
  Capacity increased from 239 to 495 [ImplSize=512]
  Capacity increased from 495 to 1007 [ImplSize=1024]
  Capacity increased from 1007 to 2031 [ImplSize=2048]
  Capacity increased from 2031 to 4079 [ImplSize=4096]
  Capacity increased from 4079 to 8175 [ImplSize=8192]
  Capacity increased from 8175 to 16367 [ImplSize=16384]
  Capacity increased from 16367 to 32751 [ImplSize=32768]
  Capacity increased from 32751 to 65519 [ImplSize=65536]
  Capacity increased from 65519 to 131055 [ImplSize=131072]
  Capacity increased from 131055 to 262127 [ImplSize=262144]
  Capacity increased from 262127 to 524271 [ImplSize=524288]
  Capacity increased from 524271 to 1048559 [ImplSize=1048576]
  Capacity increased from 1048559 to 2097135 [ImplSize=2097152]
  Capacity increased from 2097135 to 4194287 [ImplSize=4194304]
  Capacity increased from 4194287 to 8388591 [ImplSize=8388608]
  Capacity increased from 8388591 to 11534320 [ImplSize=11534337]

[[Priority=9]]
Running matrix
  Testing matrix types
  Testing whether special-case transformations match matrix multiplication

[[Priority=10]]
Running path_allocation_strategy
  Capacity increased from 0 to 54 [ImplSize=1022]
  Capacity increased from 54 to 234 [ImplSize=4082]
  Capacity increased from 234 to 475 [ImplSize=8179]
  Capacity increased from 475 to 957 [ImplSize=16373]
  Capacity increased from 957 to 1921 [ImplSize=32761]
  Capacity increased from 1921 to 3848 [ImplSize=65520]
  Capacity increased from 3848 to 7704 [ImplSize=131072]
  Capacity increased from 7704 to 15414 [ImplSize=262142]
  Capacity increased from 15414 to 30834 [ImplSize=524282]
  Capacity increased from 30834 to 61674 [ImplSize=1048562]
  Capacity increased from 61674 to 123355 [ImplSize=2097139]
  Capacity increased from 123355 to 246717 [ImplSize=4194293]
  Capacity increased from 246717 to 493441 [ImplSize=8388601]
  Capacity increased from 493441 to 678566 [ImplSize=11535726]
  Capacity increased from 678566 to 933083 [ImplSize=15862515]
  Capacity increased from 933083 to 1283012 [ImplSize=21811308]

[[Priority=11]]
Running image
  Testing BLImage::create() and BLImage::makeMutable()
  Testing BLImage::create() and BLImage::convert()
  FAILED: EXPECT_TRUE(img0.equals(img1))
  SOURCE: /builddir/build/BUILD/blend2d-5a263ce51f3f880ee6c60f6345d18c3eccbe200f/src/blend2d/image_test.cpp (Line: 92)

    Start 3: bl_test_verify_mt
2/4 Test #2: bl_test_fuzzer ...................   Passed  158.95 sec
    Start 4: bl_generator
3/4 Test #3: bl_test_verify_mt ................   Passed  1364.97 sec
4/4 Test #4: bl_generator .....................   Passed  1453.86 sec

75% tests passed, 1 tests failed out of 4

Total Test time (real) = 1612.82 sec

The following tests FAILED:
	  1 - bl_test_unit (Subprocess aborted)
Errors while running CTest
```

</details>

ref: https://download.copr.fedorainfracloud.org/results/topazus/test/fedora-rawhide-s390x/07207582-blend2d/builder-live.log.gz

https://copr.fedorainfracloud.org/coprs/topazus/test/build/7207582/

**Comment by QuLogic** [comment]
I believe that issue is at https://github.com/blend2d/blend2d/blob/master/src/blend2d/pixelconverter.cpp#L545; the converter takes the last byte of the [PX]RGB32 value (I printed out `srcAI` which was 3) whereas the test fills the expected value to the MSB of a 32-bit value: https://github.com/blend2d/blend2d/blob/master/src/blend2d/image_test.cpp#L86 On a big-endian system, that would be the first byte (I'd expect `srcAI` to be 0). But I'm not sure where `alphsShift` gets set.

**Comment by kobalicek** [comment]
Thanks a lot for the report and for the PR - I will definitely check this code out.

Unfortunately I have no big endian machine to test this on so I'm not that surprised that something is broken even when I really tried to write code that should be aware of the differences.

**Comment by kobalicek** [comment]
@QuLogic You are correct here. I just looked into this and yeah the function `bl_convert_a8_from_8888` is wrong on big endian machines.

**Comment by kobalicek** [comment]
Should be fixed by https://github.com/blend2d/blend2d/commit/e4656f4317b79891c85865023039507db142e6d3

**Comment by topazus** [comment]
> Should be fixed by [e4656f4](https://github.com/blend2d/blend2d/commit/e4656f4317b79891c85865023039507db142e6d3)

Thanks for the effort, I ascertain that the commit fixed this issue.
